subprojects/core/src/main/java/org/gradle/api/internal/file/collections/ImmutableFileCollection.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file.collections;
+
+import com.google.common.collect.ImmutableSet;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.AbstractFileCollection;
+import org.gradle.api.internal.file.IdentityFileResolver;
+import org.gradle.internal.file.PathToFileResolver;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+public class ImmutableFileCollection extends AbstractFileCollection {
+    private static final String DEFAULT_DISPLAY_NAME = "immutable file collection";
+    private static final EmptyImmutableFileCollection EMPTY = new EmptyImmutableFileCollection();
+
+    private final Set<Object> files;
+    private final String displayName;
+    private final PathToFileResolver resolver;
+
+    public static ImmutableFileCollection of(Object... files) {
+        return of(Arrays.asList(files));
+    }
+
+    public static ImmutableFileCollection of(Collection<?> files) {
+        return usingResolver(new IdentityFileResolver(), files);
+    }
```

Comment:
I think what we need is three sub-implementations of an abstract `ImmutableFileCollection`:

- one that takes a `FileResolver` and a bunch of `Object`s and resolves them when needed
- another that takes a bunch of `File` objects and just returns them (this needs no resolver at all)
- empty implementation (we already have this)

The point is that the second variant can be faster because it doesn't have to go through `DeferredUtil.unpack()` and `fileNotationParser.parseNotation()` unnecessarily. There's also no need to create a resolution context etc. for this case.

Reply:
Done.  The static factory methods in this class are roughly the same (apart from taking `Iterable<>` instead of `Collection<>`).